packages/react-components/react-combobox/src/hooks/useComboboxFilter.tsx

```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
+
+import type { ComboboxProps } from '../Combobox';
+import { Option } from '../Option';
+
+type UseComboboxFilterConfig<T extends { children: React.ReactNode; value: string } | string> = {
+  /** Provides a custom filter for the option. */
+  filter?: (optionText: string, query: string) => boolean;
+
+  /** Provides a custom message to display when there are no options. */
+  noOptionsMessage?: React.ReactNode;
+
+  /** Provides a way to map an option object to a React key. By default, "value" is used. */
+  optionToReactKey?: (option: T) => string;
+
+  /** Provides a way to map an option object to a text used for search. By default, "value" is used. */
+  optionToText?: (option: T) => string;
+
+  /** Provides a custom render for the option. */
+  renderOption?: (option: T) => JSX.Element;
+};
+
+function defaultFilter(optionText: string, query: string) {
+  if (query === '') {
+    return true;
+  }
+
+  return optionText.toLowerCase().includes(query.toLowerCase());
+}
+
+function defaultToString(option: string | { value: string }) {
+  return typeof option === 'string' ? option : option.value;
+}
+
+export function useComboboxFilter<T extends { children: React.ReactNode; value: string } | string>(
+  options: T[],
+  config: UseComboboxFilterConfig<T>,
+) {
+  const {
+    filter = defaultFilter,
+    noOptionsMessage = "We couldn't find any matches.",
+    optionToReactKey = defaultToString,
+    optionToText = defaultToString,
+
+    renderOption = (option: T) => {
+      if (typeof option === 'string') {
+        return <Option key={option}>{option}</Option>;
+      }
+
+      return (
+        <Option {...option} key={optionToReactKey(option)} text={optionToText(option)} value={option.value}>
+          {option.children}
+        </Option>
+      );
+    },
+  } = config;
+  const [query, setQuery] = React.useState<string>('');
+
+  const onChange: ComboboxProps['onChange'] = ev => {
+    setQuery(ev.target.value.trim());
+  };
+
+  const filteredOptions = React.useMemo(
+    () => options.filter(option => filter(optionToText(option), query)),
+    [options, optionToText, filter, query],
+  );
+
+  if (filteredOptions.length === 0) {
+    return {
+      children: [
+        <Option key="no-results" text="">
+          {noOptionsMessage}
```

Comment:
This should probably support a renderer, it's quite optimistic that apps would be content with a string

Reply:
In offline discussion we agreed to postpone implementation of this as there is no requirement from consumers yet and will be better to propose a solution based on it. Currently there are two options: implement a render function or as a slot.